util/src/openssl/certificates.cpp

```diff
@@ -0,0 +1,261 @@
+    LOG_ERROR(OPENSSL_LOG, "Unable to create private key object");
+    return string();
+  }
+
+  if (nullptr == PEM_read_bio_PrivateKey(priv_bio.get(), reinterpret_cast<EVP_PKEY**>(&priv_key), nullptr, nullptr)) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create private key object");
+    return string();
+  }
+  UniquePKEY pub_key(EVP_PKEY_new());
+  UniqueBIO pub_bio(BIO_new(BIO_s_mem()));
+
+  if (BIO_write(pub_bio.get(), static_cast<const char*>(public_key.c_str()), public_key.size()) <= 0) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to write public key object");
+    return string();
+  }
+  if (nullptr == PEM_read_bio_PUBKEY(pub_bio.get(), reinterpret_cast<EVP_PKEY**>(&pub_key), nullptr, nullptr)) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create public key object");
+    return string();
+  }
+
+  if (OPENSSL_FAILURE == X509_set_pubkey(cert.get(), pub_key.get())) {
+    LOG_ERROR(OPENSSL_LOG, "Failed to set public key for certificate.");
+    return {};
+  }
+
+  if (SignatureAlgorithm::RSA == signingAlgo) {
+    if (OPENSSL_FAILURE == X509_sign(cert.get(), priv_key.get(), EVP_sha256())) {
+      LOG_ERROR(OPENSSL_LOG, "Failed to sign certificate using RSA private key.");
+      return {};
+    }
+  } else if (SignatureAlgorithm::EdDSA == signingAlgo) {
+    if (OPENSSL_FAILURE == X509_sign(cert.get(), priv_key.get(), nullptr)) {
+      LOG_ERROR(OPENSSL_LOG, "Failed to sign certificate using EdDSA private key.");
+      return {};
+    }
+  }
+
+  UniqueBIO outbio(BIO_new(BIO_s_mem()));
+  if (OPENSSL_FAILURE == PEM_write_bio_X509(outbio.get(), cert.get())) {
+    LOG_ERROR(OPENSSL_LOG, "Unable to create certificate object");
+    return string();
+  }
+  string certStr;
+  int certLen = BIO_pending(outbio.get());
+  certStr.resize(certLen);
+  const auto res = BIO_read(outbio.get(), (void*)&(certStr.front()), certLen);
+  if (OPENSSL_FAILURE == res || OPENSSL_ERROR == res) {
+    LOG_ERROR(OPENSSL_LOG, "Failed to read data from the BIO certifiate object.");
+    return {};
+  }
+  return certStr;
+}
+
+bool verifyCertificate(X509& cert, const string& public_key) {
+  UniquePKEY pub_key(EVP_PKEY_new());
+  UniqueBIO pub_bio(BIO_new(BIO_s_mem()));
+
+  if (BIO_write(pub_bio.get(), static_cast<const char*>(public_key.c_str()), public_key.size()) <= 0) {
+    return false;
+  }
+  if (nullptr == PEM_read_bio_PUBKEY(pub_bio.get(), reinterpret_cast<EVP_PKEY**>(&pub_key), nullptr, nullptr)) {
+    return false;
+  }
+  return (OPENSSL_SUCCESS == X509_verify(&cert, pub_key.get()));
+}
+
+bool verifyCertificate(const X509& cert_to_verify,
+                       const string& cert_root_directory,
+                       uint32_t& remote_peer_id,
+                       string& conn_type,
+                       bool use_unified_certs) {
+  // First get the source ID
+  static constexpr size_t SIZE = 512;
+  string subject(SIZE, 0);
+  X509_NAME_oneline(X509_get_subject_name(&cert_to_verify), subject.data(), SIZE);
+
+  int peerIdPrefixLength = 3;
```

Comment:
A general comment about coding style:
Why some of the code is snake-case and some camel-case?
I know that if I modify an existing code i tend to keep the existing style.
But this is a new file and it mixes between 2 styles...looks strange to me.
e.g peerIdPrefixLengthand vs origin_cert_path

Reply:
Resolved. https://github.com/vmware/concord-bft/pull/2851